transformer_engine/pytorch/attention.py

```diff
@@ -1062,37 +1106,78 @@ def _rotate_half(x: torch.Tensor) -> torch.Tensor:
+    transpose_output_memory: bool, default = False
+        Whether to transpose the 's' and 'b' dimension of the output's underlying memory format.
+        This is very helpful when you want to get a contiguous tensor after calling
+        `output.transpose(0, 1)`. It's only supported when `fused` is True and
+        `tensor_format` is 'sbhd' or 'bshd'.
+    cu_seqlens: torch.Tensor, default = None.
+        Cumulative sum of sequence lengths in a batch for `t`, with shape [b + 1] and
+        dtype torch.int32. Only valid when `tensor_format` is 'thd'.
     """
+    if fused:
+        assert not (
+            transpose_output_memory and tensor_format == "thd"
+        ), "transpose_output_memory only supports 'sbhd' or 'bshd' format."
+        assert (
+            tensor_format != "thd" or cu_seqlens is not None
+        ), "cu_seqlens must not be None when tensor_format is 'thd'."
+        return FusedRoPEFunc.apply(
+            t, freqs, transpose_output_memory, tensor_format, cu_seqlens
+        )
+
+    assert tensor_format in ("sbhd", "bshd"), (
+        "Only formats `sbhd` or `bshd` are supported for input tensor `t` "
+        f"when fused is False, got {tensor_format}."
+    )
+    assert (
+        not transpose_output_memory
+    ), "transpose_output_memory is not supported when fused is False."
+
+    max_seq_len = freqs.shape[0]
+    cur_seq_len = t.shape[1] if tensor_format == "bshd" else t.shape[0]
+
+    # Only apply the rotary embeddings up to the sequence length of the running
+    # input.
+    assert cur_seq_len <= max_seq_len, (
+        f"Rotary Embeddings only supported up to {max_seq_len} sequence length!"
+    )
+    freqs = freqs[:cur_seq_len]
+    if tensor_format == "bshd":
+        freqs = freqs.transpose(0, 1)  # [seq, 1, 1, dim] -> [1, seq, 1, dim]
+    # cos/sin first then dtype conversion for better precision
     cos_ = torch.cos(freqs).to(t.dtype)
     sin_ = torch.sin(freqs).to(t.dtype)
 
-    if fused:
-        return FusedRoPEFunc.apply(t, cos_, sin_)
-    else:
-        rot_dim = freqs.shape[-1]
-        # ideally t_pass is empty so rotary pos embedding is applied to all tensor t
-        t, t_pass = t[..., :rot_dim], t[..., rot_dim:]
-
-        # first part is cosine component
-        # second part is sine component, need to change signs with _rotate_half method
-        cos_ = torch.cos(freqs).to(t.dtype)
-        sin_ = torch.sin(freqs).to(t.dtype)
-        t = (t * cos_) + (_rotate_half(t) * sin_)
-        return torch.cat((t, t_pass), dim=-1)
+    rot_dim = freqs.shape[-1]
+    # ideally t_pass is empty so rotary pos embedding is applied to all tensor t
+    t, t_pass = t[..., :rot_dim], t[..., rot_dim:]
+
+    # first part is cosine component
+    # second part is sine component, need to change signs with _rotate_half method
+    t = (t * cos_) + (_rotate_half(t) * sin_)
+    return torch.cat((t, t_pass), dim=-1)
```

Comment:
To return a BSHD tensor:
```suggestion
    emb = torch.cat((t, t_pass), dim=-1)
    if tensor_format == "bshd":
        emb = emb.transpose(0, 1).contiguous()
    return emb
```

Reply:
I think `tensor_format` is used to indicate the format of input tensor. When `tensor_format == "bshd"`, `t` should already be in `bshd` and `freqs` has been tranposed. So Line 1179 is done in `bshd`, and we don't need the transpose here.